design/design-doc-agent.md

```diff
@@ -0,0 +1,586 @@
+team, which is `*Requirement` types and consume the infrastructure by mounting
+the secret whose name they specified on the `*Requirement` custom resource. To
+keep this experience, we need to have a synchronization loop for the following
+resources:
+
+* Pull
+  * `CustomResourceDefinition`s of all types that we want the applications to be
+    able to manage and view:
+    * Requirements that are published via `InfrastructurePublication`s. The
+      source of truth will be the remote cluster.
+    * `InfrastructureDefinition`, `InfrastructurePublication` and `Composition`.
+  * `Composition`s to discover what's available to choose. CRs of this type will
+    be read-only, and the source of truth will be the central cluster.
+  * `InfrastructurePublication`s to discover what's availabe as published API.
+    Read-only.
+  * `InfrastructureDefinition`s to discover how the secret keys are shaped.
+    Read-only.
+  * `Secret`s that are result of the infrastructure that is provisioned so that
+    it can be mounted to `Pod`s. Read-only.
+* Push
+  * `*Requirement` custom resources so that infrastructure can be requested.
+    Read and write permissions in a specific namespace in the central cluster will
+    be needed.
+
+### RBAC
+
+As we have two different Kubernetes cluster, there will be two separate security
+domains and because of that, the `ServiceAccount`s in the remote cluster will
+not be able to do any operation in the central cluster. Since the entity that
+will execute the operations in the central cluster is the agent, we need to
+define how we can deliver the necessary credentials to the agent so that it can
+connect to the central cluster. Additionally, it will need some permissions to
+execute operations in the remote cluster like `Secret` and
+`CustomResourceDefinition` creation. We will look at how the agent will be
+granted permissions to do its job in two separate domains with different
+mechanisms.
+
+#### Authenticating to The Central Cluster
+
+In order to execute any operation, a `ServiceAccount` needs to exist in the
+central cluster with appropriate permissions to read during pull and write
+during push operations while synchronizing with central cluster. Since the agent
+is running in the remote cluster, the credentials of this `ServiceAccount` will
+be stored in a `Secret` in the remote cluster. Alongside the credentials, the
+agent needs to know the namespace that it should sync to in the central cluster.
+
+The way to have the agent find the right `Secret` to use in order to
+authenticate to the central cluster for synchronization of a given requirement
+will consist of two steps:
+
+* Look for specific annotations on the requirement type to specify what `Secret`
+  needs to be used:
+  * `agent.crossplane.io/credentials-secret-name: sa-secret` for the full
+    kubeconfig to connect to the central cluster.
+  * `agent.crossplane.io/namespace: my-app-1` for the namespace it should sync
```

Comment:
> Is this annotation on the *Requirement CRD or is this added on a *Requirement CR to effect it's sync behavior?

It's on the `*Requirement` custom resource, not on its definition.

> I think this annotation is based on explicit agent configuration. i.e. the CR is written to a agent cluster namespace, and the agent resolves the remote namespace for the local, defaulting to the same name if no-override is configured.

Do you mean if I create a requirement in namespace `foo` in my remote cluster, then by default, it should sync to namespace `foo` in the central cluster? I'd be wary of such pairing because the identification space of the central cluster is made up of several remote clusters and their namespaces, so, default wouldn't really be useful. For example, I can see the case where several remote clusters just using `default` namespace initially and they'd conflict by default. So, that exact pairing may start to not work well really quickly.

Reply:
If I'm setting this agent up as an infra-owner defining the mapping seems totally reasonable as well. If we allow arbitrary cross mapping via annotations, it seems like it could get very confusing, very quickly. I might argue that we enforce a ns:ns mapping so we enforce a bit of organization on how things are provisioned and avoid maybe opening us up to some odd security issues.

It seems that sa-secret has to be a property of every local namespace and there's no way for infra-owner to set a default at the infra level that works for a set of namespaces which is what I was imagining.

This is what I was imaging not being a good behavior that we would by default not want to happen:
![RequirementSync](https://user-images.githubusercontent.com/195363/86264580-8ed44600-bb77-11ea-8cf0-5828c2c8caa8.png)

Do you see as a use case an infra-owner who wants to setup a set of app namespaces to connect to a set of crossplane namespaces without needing separate service accounts for each namespace? I was kind of imaging, as an infra owner I want to provision an app cluster, and then install the agent on it to connect back to crossplane for all namespaces.

Reply:
> something less flexible and more predictable would be ideal

@muvaf Pushed a change to this PR since I left this comment that aligns pretty closely with what I was thinking. The relationship between remote and central resources are now defined by annotations at the namespace level rather than annotations on the CRs themselves. I suspect this is about the right level of flexibility and granularity.